File: tier1_foundation/string/problems/p520_detect_capital.mjs
/**
 * P520: Detect Capital [PREMIUM] (Easy)
 * https://leetcode.com/problems/detect-capital/
 * Topics: String
 * 
 * We define the usage of capitals in a word to be right when one of the following cases holds:
 * Given a string word, return true if the usage of capitals in it is right.
 * Example 1:
 *     Input: word = "USA"
 *     Output: true
 * 
 * Example 2:
 *     Input: word = "FlaG"
 *     Output: false
 * 
 * Constraints:
 *     - 1 <= word.length <= 100
 *     - word consists of lowercase and uppercase English letters.
 * 
 * Template (python3):
 *     class Solution:
 *         def detectCapitalUse(self, word: str) -> bool:
 */
function solve(word)
  return word === word.toUpperCase() || word === word.toLowerCase() || word.slice(1) === word.slice(1).toLowerCase(); {
    throw new Error("NotImplementedError");
}

const tests = [
  { label: "example 1", input: ["USA"], expected: true },
  { label: "example 2", input: ["FlaG"], expected: false },
  { label: "single lowercase", input: ["a"], expected: true },
  { label: "single uppercase", input: ["Z"], expected: true },
  { label: "all lowercase", input: ["leetcode"], expected: true },
  { label: "title case", input: ["Google"], expected: true },
  { label: "lowercase then uppercase", input: ["mL"], expected: false },
];
let passed = 0;
for (let i = 0; i < tests.length; i++) {
  const t = tests[i];
  const got = solve(t.input[0]);
  if (JSON.stringify(got) === JSON.stringify(t.expected)) {
    passed++;
    console.log(`  Test ${i + 1} (${t.label}): PASS`);
  } else {
    console.log(`  Test ${i + 1} (${t.label}): FAIL`);
    console.log(`    Expected: ${JSON.stringify(t.expected)}\n    Got:      ${JSON.stringify(got)}`);
  }
}
console.log(`\n  ${passed}/${tests.length} passed`);
process.exit(passed === tests.length ? 0 : 1);

File: tier1_foundation/string/solutions/p520_detect_capital.c
/*
 * P520: Detect Capital [PREMIUM] (Easy)
 * https://leetcode.com/problems/detect-capital/
 * Topics: String
 *
 * We define the usage of capitals in a word to be right when one of the following cases holds:
 * Given a string word, return true if the usage of capitals in it is right.
 * Example 1:
 *     Input: word = "USA"
 *     Output: true
 *
 * Example 2:
 *     Input: word = "FlaG"
 *     Output: false
 *
 * Constraints:
 *     - 1 <= word.length <= 100
 *     - word consists of lowercase and uppercase English letters.
 *
 * Template (python3):
 *     class Solution:
 *         def detectCapitalUse(self, word: str) -> bool:
 */


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

static int detectCapitalUse(const char *word)
{
    int upper = 0, first_upper = isupper((unsigned char)word[0]);
    for (int i = 0; word[i]; i++)
        if (isupper((unsigned char)word[i]))
            upper++;
    int len = (int)strlen(word);
    return upper == len || upper == 0 || (first_upper && upper == 1);
}

int main(void)
{
    struct { const char *label; const char *input; int expected; } tests[] = {
        {"example 1",       "USA",      1},
        {"example 2",       "FlaG",     0},
        {"single lowercase","a",        1},
        {"single uppercase","Z",        1},
        {"all lowercase",   "leetcode", 1},
        {"title case",      "Google",   1},
        {"lowercase then uppercase", "mL", 0},
    };
    int n = (int)(sizeof(tests) / sizeof(tests[0]));
    int passed = 0;
    for (int i = 0; i < n; i++) {
        int got = detectCapitalUse(tests[i].input);
        if (got == tests[i].expected) {
            printf("  Test %d (%s): PASS\n", i + 1, tests[i].label);
            passed++;
        } else {
            printf("  Test %d (%s): FAIL\n", i + 1, tests[i].label);
            printf("    Expected: %d\n    Got: %d\n", tests[i].expected, got);
        }
    }
    printf("\n  %d/%d passed\n", passed, n);
    return passed == n ? 0 : 1;
}
